hkmc2/shared/src/main/scala/hkmc2/codegen/Block.scala

```diff
@@ -89,10 +93,13 @@ case class TryBlock(sub: Block, finallyDo: Block, rest: Block) extends Block wit
 case class Assign(lhs: Local, rhs: Result, rest: Block) extends Block with ProductWithTail
 // case class Assign(lhs: Path, rhs: Result, rest: Block) extends Block with ProductWithTail
 
-case class AssignField(lhs: Path, nme: Tree.Ident, rhs: Result, rest: Block)(symbol: Opt[FieldSymbol]) extends Block with ProductWithTail
+case class AssignField(lhs: Path, nme: Tree.Ident, rhs: Result, rest: Block)(val symbol: Opt[FieldSymbol]) extends Block with ProductWithTail
 
 case class Define(defn: Defn, rest: Block) extends Block with ProductWithTail
 
+case class HandleBlock(lhs: Local, res: Local, cls: Path, handlers: Ls[Handler], body: Block, rest: Block) extends Block with ProductWithTail
```

Comment:
So the fact that the result has to be stored in `res` seems to imply that no calls inside `body` are considered tail calls, right? But we might want to optimize such tail calls by not instrumenting them (when `rest` is empty, that is), so it would be more helpful if they were apparent and not hidden behind `Assign` nodes.

Reply:
Optimizating tail call inside handle block would require verifying all handlers are tail-resumptive or else the tail call may not be a tail call in the result. The handle block is responsible for resuming the non tail-resumptive handlers.

Reply:
Ah, that's right. But that's an easy check to make, and it's probably a relatively common case. So it would still make sense if we could optimize it.

Would it be difficult to express the handled code as a block with possible tail calls?

Reply:
Ok actually I guess the optimization wouldn't be sound because the resuming code might use the effect of the handler in question, so we must register the fact that this part of the code has this handler in the continuation object. Is that handled correctly in your current implementation of the instrumentation?

Reply:
The tail position in the handle block are marked with HandleBlockReturn currently.

Actually I think even if all handlers are tail calls, we still need to check whether the returned value is a continuation so I think it can't be a tail call. But the common scenario should not make a handler for each recursive call anyway, and the recursive function should not contain any handle block

Reply:
The current transformation inserts code to the handle block before adding continuation checks. So it should be fine because continuation check is only skipped when it is tail after adding the handle block code

Reply:
Yeah, that's a good point.